prow/cmd/deck/rerun.go

```diff
@@ -18,21 +18,159 @@ package main
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
 	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	v1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
 	prowv1 "k8s.io/test-infra/prow/client/clientset/versioned/typed/prowjobs/v1"
 	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/gerrit/client"
+	"k8s.io/test-infra/prow/github"
 	"k8s.io/test-infra/prow/githuboauth"
 	"k8s.io/test-infra/prow/pjutil"
 	"k8s.io/test-infra/prow/plugins"
 	"k8s.io/test-infra/prow/plugins/trigger"
 )
 
+var (
+	// Stores the annotations and labels that are generated
+	// and specified within components.
+	ComponentSpecifiedAnnotationsAndLabels = sets.NewString(
+		// Labels
+		client.GerritRevision,
+		client.GerritPatchset,
+		client.GerritReportLabel,
+		github.EventGUID,
+		"created-by-tide",
+		// Annotations
+		client.GerritID,
+		client.GerritInstance,
+	)
+)
+
+func verifyRerunRefs(refs *prowapi.Refs) error {
+	if refs == nil {
+		return errors.New("Refs must be supplied")
+	}
+	if len(refs.Org) == 0 {
+		return errors.New("org must be supplied")
+	}
+	if len(refs.Repo) == 0 {
+		return errors.New("repo must be supplied")
+	}
+	if len(refs.BaseRef) == 0 {
+		return errors.New("baseRef must be supplied")
+	}
+	return nil
+}
+
+func setRerunOrgRepo(refs *prowapi.Refs, labels map[string]string) string {
+	org, repo := refs.Org, refs.Repo
+	orgRepo := org + "/" + repo
+	// Add "https://" prefix to orgRepo if this is a gerrit job.
+	// (Unfortunately gerrit jobs use the full repo URL as the identifier.)
+	prefix := "https://"
+	if labels[client.GerritRevision] != "" && !strings.HasPrefix(orgRepo, prefix) {
+		orgRepo = prefix + orgRepo
+	}
+	return orgRepo
+}
+
+func getPresubmitSpec(cfg config.Getter, name string, refs *prowapi.Refs, labels map[string]string) (*v1.ProwJobSpec, map[string]string, map[string]string, error) {
+	// presubmit jobs require Refs
+	if err := verifyRerunRefs(refs); err != nil {
+		return nil, nil, nil, err
+	}
+	var presubmitJob *config.Presubmit
+	branch := refs.BaseRef
+	orgRepo := setRerunOrgRepo(refs, labels)
+	// Get presubmits from Config alone.
+	presubmits := cfg().GetPresubmitsStatic(orgRepo)
+	for _, job := range presubmits {
+		job := job
+		if !job.CouldRun(branch) { // filter out jobs that are not branch matching
+			continue
+		}
+		if job.Name == name {
+			if presubmitJob != nil {
+				return nil, nil, nil, fmt.Errorf("%s matches multiple prow jobs from orgRepo %q", name, orgRepo)
+			}
+			presubmitJob = &job
+		}
+	}
+	if presubmitJob == nil {
+		return nil, nil, nil, fmt.Errorf("failed to find associated presubmit job %q from orgRepo %q", name, orgRepo)
+	}
+
+	prowJobSpec := pjutil.PresubmitSpec(*presubmitJob, *refs)
+	return &prowJobSpec, presubmitJob.Labels, presubmitJob.Annotations, nil
+}
+
+func getPostsubmitSpec(cfg config.Getter, name string, refs *prowapi.Refs, labels map[string]string) (*v1.ProwJobSpec, map[string]string, map[string]string, error) {
```

Comment:
This is almost identical to getPresubmitSpec - Maybe use generics to dedupe?

Reply:
I previously attempted to use generics, I faced some errors where the compiler couldn't determine which type to use when calling `job.CouldRun`. Both Presubmit and Postsubmit have this function. Do you know a work around and is it worth implementing?

Reply:
Something like this at least compiles:
```diff
diff --git a/prow/cmd/deck/rerun.go b/prow/cmd/deck/rerun.go
index 2e5e061d86..a86a5dda10 100644
--- a/prow/cmd/deck/rerun.go
+++ b/prow/cmd/deck/rerun.go
@@ -84,61 +84,48 @@ func setRerunOrgRepo(refs *prowapi.Refs, labels map[string]string) string {
 }
 
 func getPresubmitSpec(cfg config.Getter, name string, refs *prowapi.Refs, labels map[string]string) (*v1.ProwJobSpec, map[string]string, map[string]string, error) {
-	// presubmit jobs require Refs
-	if err := verifyRerunRefs(refs); err != nil {
-		return nil, nil, nil, err
-	}
-	var presubmitJob *config.Presubmit
-	branch := refs.BaseRef
-	orgRepo := setRerunOrgRepo(refs, labels)
-	// Get presubmits from Config alone.
-	presubmits := cfg().GetPresubmitsStatic(orgRepo)
-	for _, job := range presubmits {
-		job := job
-		if !job.CouldRun(branch) { // filter out jobs that are not branch matching
-			continue
-		}
-		if job.Name == name {
-			if presubmitJob != nil {
-				return nil, nil, nil, fmt.Errorf("%s matches multiple prow jobs from orgRepo %q", name, orgRepo)
-			}
-			presubmitJob = &job
-		}
-	}
-	if presubmitJob == nil {
-		return nil, nil, nil, fmt.Errorf("failed to find associated presubmit job %q from orgRepo %q", name, orgRepo)
-	}
-
-	prowJobSpec := pjutil.PresubmitSpec(*presubmitJob, *refs)
-	return &prowJobSpec, presubmitJob.Labels, presubmitJob.Annotations, nil
+	return getPreOrPostSpec(cfg().GetPresubmitsStatic, pjutil.PresubmitSpec, name, refs, labels)
 }
 
 func getPostsubmitSpec(cfg config.Getter, name string, refs *prowapi.Refs, labels map[string]string) (*v1.ProwJobSpec, map[string]string, map[string]string, error) {
+	return getPreOrPostSpec(cfg().GetPostsubmitsStatic, pjutil.PostsubmitSpec, name, refs, labels)
+}
+
+type preOrPostsubmit interface {
+	GetName() string
+	CouldRun(string) bool
+	GetLabels() map[string]string
+	GetAnnotations() map[string]string
+}
+
+func getPreOrPostSpec[p preOrPostsubmit](jobGetter func(string) []p, creator func(p, v1.Refs) v1.ProwJobSpec, name string, refs *prowapi.Refs, labels map[string]string) (*v1.ProwJobSpec, map[string]string, map[string]string, error) {
 	// postsubmit jobs require Refs to be set
 	if err := verifyRerunRefs(refs); err != nil {
 		return nil, nil, nil, err
 	}
-	var postsubmitJob *config.Postsubmit
+	var result *p
+
 	branch := refs.BaseRef
 	orgRepo := setRerunOrgRepo(refs, labels)
-	postsubmits := cfg().GetPostsubmitsStatic(orgRepo)
-	for _, job := range postsubmits {
+	for _, job := range jobGetter(orgRepo) {
 		job := job
-		if !job.CouldRun(branch) { // filter out jobs that are not branch matching
+		if !job.CouldRun(branch) {
 			continue
 		}
-		if job.Name == name {
-			if postsubmitJob != nil {
+		if job.GetName() == name {
+			if result != nil {
 				return nil, nil, nil, fmt.Errorf("%s matches multiple prow jobs from orgRepo %q", name, orgRepo)
 			}
-			postsubmitJob = &job
+			result = &job
 		}
 	}
-	if postsubmitJob == nil {
-		return nil, nil, nil, fmt.Errorf("failed to find associated postsubmit job %q from orgRepo %q", name, orgRepo)
+
+	if result == nil {
+		return nil, nil, nil, fmt.Errorf("failed to find job for %s for orgRepo %q", name, orgRepo)
 	}
-	prowJobSpec := pjutil.PostsubmitSpec(*postsubmitJob, *refs)
-	return &prowJobSpec, postsubmitJob.Labels, postsubmitJob.Annotations, nil
+
+	prowJobSpec := creator(*result, *refs)
+	return &prowJobSpec, (*result).GetLabels(), (*result).GetAnnotations(), nil
 }
 
 func getPeriodicSpec(cfg config.Getter, name string) (*v1.ProwJobSpec, map[string]string, map[string]string, error) {
diff --git a/prow/config/jobs.go b/prow/config/jobs.go
index 02e50a68c4..2a9bf643d1 100644
--- a/prow/config/jobs.go
+++ b/prow/config/jobs.go
@@ -136,6 +136,18 @@ type JobBase struct {
 	UtilityConfig
 }
 
+func (jb JobBase) GetName() string {
+	return jb.Name
+}
+
+func (jb JobBase) GetLabels() map[string]string {
+	return jb.Labels
+}
+
+func (jb JobBase) GetAnnotations() map[string]string {
+	return jb.Annotations
+}
+
 // +k8s:deepcopy-gen=true
 
 // Presubmit runs on PRs.

```

Reply:
Thanks Alvaro for this code suggestion, I have taken it up. I'm not too familiar with Go's generics, yet, but this implementation seems to work as desired.